### DISCUSSION TITLE
Custom topology keys for CSI nodes

### DISCUSSION THREAD

**Discussion by Kavinraja-G** [topolvm/topolvm#712] [Q&A]
I have configured the `allowedTopologies` for the StorageClasses created by the topolvm. In order to make it work, I need to add those custom topologyKeys in the `CSINodes` as well.

By deafult, it adds the hostName to the Keys, but I need to add the customKeys/Labesl to them. Is there any workaround to achieve this?

Scenario is I have certains nodes with custom labels, say `nodePoolType: stateful` which has special disks that are to be consumed by topolvm.

**Comment by daichimukai** [comment]
Hi @Kavinraja-G,

> Scenario is I have certains nodes with custom labels, say nodePoolType: stateful which has special disks that are to be consumed by topolvm.

We are using TopoLVM in a similar environment. In your words, we just deploy topolvm-node (and lvmd) to the nodes with the label `nodePoolType: stateful`. 

> I have configured the allowedTopologies for the StorageClasses created by the topolvm. In order to make it work, I need to add those custom topologyKeys in the CSINodes as well.

We don't set `allowedTopoloogies` for the storage classes. CSINodes are automatically configured by the sidecar container (node-driver-registrarer) of topolvm-node. This setup is working well for us. 

I hope this will help.

**Reply by Kavinraja-G** [reply]
@daichimukai Thanks for the response, let me try that real quick. I remember seeing errors related to mismatch in the topology map when the controller tries to provision the PV.

And, did you set the nodeTopology in the CSIStorageCapacity as well? Something like this --
```yaml
nodeTopology:
  matchLabels:
    nodePoolType: stateful
```

**Reply by Kavinraja-G** [reply]
I got these warnings in csi-provisioner container but it got finally provisioned. Weird :/
```
'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "topolvm-provisioner": error generating accessibility requirements: no topology key found on CSINode ip-x-x-y-y
```

Do I need to worry about this @daichimukai?

**Reply by daichimukai** [reply]
> And, did you set the nodeTopology in the CSIStorageCapacity as well?

We have no CSIStorageCapacity resource in our cluster. Are you using storage capacity tracking? I'm not confident as we are using scheduler extender, but you might need `allowedTopologies` in storage class or something similar.


> I got these warnings in csi-provisioner container but it got finally provisioned. Weird :/

Hmm, I've never seen this message.

**Reply by Kavinraja-G** [reply]
I'm not using `scheduler-extender`, since I'm deploying topolvm in a managed cluster. So I enabled `storageCapacityTracking`. And, without any `allowedTopologies` in storageClass I got the above warnings for few retries in csi-provisioner container and it was able to finally provision the PV.

**Reply by daichimukai** [reply]
Do you continue to get that error if you set `allowedTopologies` in the storage class?

**Reply by Kavinraja-G** [reply]
No I didn't tried after that. Without `allowedTopologies` in StorageClass I got errors few times and it got finally provisioned pv.

**Reply by daichimukai** [reply]
I see. IMHO, there is no need to worry about that errors if your volumes are provisioned on the intended nodes and being able to mount normally. If you are concerned about the errors, please try to configure `allowedTopologies` in the storage class. Thanks.

**Reply by Kavinraja-G** [reply]
Hey @daichimukai  I did set the `StorageCapacities`, but it can't able to scale up the nodes when I request for new pods. (It was able to scale-up from 0 for the first time)

In the pod, I'm seeing `running PreBind plugin "VolumeBinding": binding volumes: provisioning failed for PVC "redpanda-dev-0"`

In the `csi-provisioner` container logs, I'm seeing the below errors --
```bash
I0705 09:37:21.744515       1 event.go:285] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"sandbox", Name:"datadir-sbx-dev-2", UID:"asdaqdwsc-asdqw32-sa3d-asd34-asd12eg", APIVersion:"v1", ResourceVersion:"707992613", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "topolvm-provisioner": rpc error: code = ResourceExhausted desc = no enough space left on VG: free=289386004480, requested=536870912000
```

@peng225 any idea on the above issue ?

**Reply by daichimukai** [reply]
It seems the node does not have enough VG capacity.

> no enough space left on VG: free=289386004480, requested=536870912000

**Reply by Kavinraja-G** [reply]
Yeah @daichimukai Doesn't it request for a new node via `cluster-autoscaler` ? We have cluster autoscaler enabled but the pod is still saying the `preBind` error and it is not scaling up.

**Reply by Kavinraja-G** [reply]
@daichimukai any idea about this issue ?

**Reply by daichimukai** [reply]
@Kavinraja-G 
It seems that cluster autoscaler would not work well for the PVCs with binding mode `WaitForFirstConsumer`. This may be a reason that your cluster did not scale up.
cf. https://github.com/kubernetes/autoscaler/issues/1658

(Thanks to @toshipp for information)

**Reply by Kavinraja-G** [reply]
@daichimukai @toshipp I don't think so, because I was able to scale-up the nodes from `0` with `cluster-autoscaler`. But beyond that topolvm is failing with `preBind error & no enough space left on VG`. 

topolvm are dynamic in provisioning the volumes it should work with `CSIStorageCapacity` as desribed here right ?

**Reply by daichimukai** [reply]
> I was able to scale-up the nodes from 0 with cluster-autoscaler.

Isn't it just scaled up with the creation of Pods that don't use PVCs with `WaitForFirstConsumer`?

Anyway, TopoLVM does nothing for cluster autoscaler, so I think it is an issue of cluster autoscaler that your cluster does not scale up.

**Reply by Kavinraja-G** [reply]
@daichimukai Can you help me understanding the use of `CSIStorageCapacity` or StorageCapacityTracking over the scheduler extender?

**Reply by daichimukai** [reply]
As far as I know, topolvm-scheduler (scheduler extender for TopoLVM) does nothing with CSIStorageCapacity.

**Reply by dee-kryvenko** [reply]
@daichimukai the CSIStorageCapacity that topolvm creates for individual nodes when `storageCapacityTracking` is enabled has this:

```
nodeTopology:
  matchLabels:
    topology.topolvm.io/node: ip-10-221-135-172
```

It does not include any other labels that the StorageClass may possibly have, which are required for Cluster Autoscaler to be able to scale nodepools from 0.

As a result, topolvm fails to provision a volume with:

```
  Warning  ProvisioningFailed    3m13s                 topolvm.io_topolvm-operator-controller-b7dfb99fc-jlzsj_2e24b7f3-b6f1-4baf-826b-3c0c1a96714c  failed to provision volume with StorageClass "my-sc": error generating accessibility requirements: topology map[topology.topolvm.io/node:ip-10-221-135-172] from selected node "ip-10-221-135-172" is not in requisite: [map[foo:bar] map[foo:baz]]
```

A storage class looks something like:

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: my-sc
parameters:
  csi.storage.k8s.io/fstype: ext4
  topolvm.io/device-class: ssd-thin
provisioner: topolvm.io
reclaimPolicy: Delete
allowVolumeExpansion: true
volumeBindingMode: WaitForFirstConsumer
allowedTopologies:
- matchLabelExpressions:
  - key: foo
    values:
    - bar
    - baz
```

What the topolvm users need is a configuration that would allow to customize what goes into `nodeTopology.matchLabels` on the CSIStorageCapacity besides `topology.topolvm.io/node`.

**Reply by dee-kryvenko** [reply]
A similar feature: https://github.com/openebs/lvm-localpv/blob/c97c4a72178d120390e1c77bd2269181ce37d049/docs/faq.md#1-how-to-add-custom-topology-key

**Reply by daichimukai** [reply]
Thank you for information. If you want that feature, please open an issue.

**Comment by peng225** [comment]
This discussion has been closed due to inactivity. Please feel free to reopen this discussion (or open a new one) if you still need to discuss this topic. Thank you for your contribution.
